File: advent7/advent7.dart
import 'dart:io';
List paths = [];
var map = <String,directory> {};
class directory {
	List<String> children;
	String parent;
	List<file> files;
	String name;
	directory(this.children,this.parent,this.files,this.name);
	num get_size(){
		num sum = 0;
		for ( file i in files ){
			sum+=i.size;
		}
		for ( String i in children ){
			sum+=map[this.name + i + '/']!.get_size();
		}
		return sum;
	}
}
class file {
	String name ;
	int size ;
	file(this.name,this.size);
}
main(){
	String curpath = '/';
	directory home = directory([],'',[],'/');
	map.putIfAbsent('/',(){return home;});
	var line = stdin.readLineSync();
	while (line!='' && line!=null){
		line = stdin.readLineSync();
		if ( line == '' || line==null ) break;
		/* print(line); */
		if (line.contains('cd ..')){
			/* print(curpath); */
			curpath = curpath.substring(0,curpath.length - 1);
			curpath = curpath.substring(0,curpath.lastIndexOf('/')+1);
			/* print(curpath); */
		}
		else if (line.contains('cd ')){
			String sub = line.substring(line.indexOf('cd ') + 3);
			curpath += sub + '/' ;
		}
		else if (line.contains('ls ')){}
		else if(line.contains('dir ')) {
			String name = line.substring(line.indexOf('dir ') + 4 );
			directory dir = directory([],curpath,[],curpath+name+'/');
			map.putIfAbsent(curpath+name+'/',(){return dir;});
			paths.add(curpath+name+'/');
			map[curpath]!.children.add(name);
		}
		else if (!line.contains('\$')){
			int size = int.parse(line.substring(0,line.indexOf(' ')));
			String name = line.substring(line.indexOf(' ')+1);
			map[curpath]!.files.add(file(name,size));
		}
	}
	num sum = 0;
	for (var i in paths){
		num n = map[i]!.get_size();
		if ( n <= 100000 ) {
			sum+=n;
		}
	}
	print(sum);
}

File: advent7/advent7.py
directories:dict = {}
curpath = '/'

class file:
    name: str
    size: int
    def __init__(self,name,size):
        self.name = name 
        self.size = size




class directory: 
    files:list[file]
    children: list[str]
    pathname: str
    def __init__(self,pathname):
        self.pathname = pathname
        self.children = []
        self.files = []

    def get_size(self):
        sum = 0
        for i in self.files:
            sum+= i.size
        for i in self.children:
            newpath = self.pathname + i + '/'
            sum+= directories[newpath].get_size()
        return sum


home = directory('/')
directories[home.pathname] = home
cur = home
with open('input.txt','r',encoding='utf-8') as r:

    for i in r:
        
        if '$' in i and 'cd /\n' in i:
            curpath ='/'
    
        elif '$' in i and 'cd ..' in i:
            if i.count('/') == 2:
                curpath = '/'
            curpath = curpath[::-1]
            curpath = curpath[1:]
            curpath = curpath[curpath.index('/'):]
            curpath = curpath[::-1]
            
        elif '$' in i and 'cd ' in i:
            
            curpath += i[i.index('cd ')+3:i.index('\n')] + '/'
            
        elif '$' in i and 'ls' in i:
            None
            print(curpath)
        elif 'dir ' in i :
            directories[curpath + i[i.index('dir ') + 4:i.index('\n')] + '/'] = directory(curpath+i[i.index('dir ' )+ 4:i.index('\n')] + '/')
            print(directories[curpath])
            print(directories)
            directories[curpath].children.append(i[i.index('dir ') + 4 :i.index('\n')])
        else:
            newfilename = i[i.index(' ')+1:i.index('\n')]
            newfilesize = int(i[:i.index(' ')])
            directories[curpath].files.append(file(newfilename,newfilesize))

s = 0
for i in directories:
    x = directories[i].get_size()
    if x <= 100000 :
        s+= x
print(s)

         

used = directories['/'].get_size()
available = 70000000 - used 
print (available)
needed = 30000000 - available 
smallestOverNeeded = directories['/'].get_size()
for i in directories: 
    if directories[i].get_size() > needed and directories[i].get_size() < smallestOverNeeded :
        smallestOverNeeded = directories[i].get_size()
print (smallestOverNeeded)


